### ISSUE TITLE
Gradient of constant evaluates to NaN

### ISSUE THREAD

**Issue by xeray** [Theano/Theano#3006]
Hello,

I have observed that taking the gradient of a constant w.r.t to parameters gives NaN for the gradient of those parameters, where I expect it to be zero. I know, taking the gradient of a constant is sort of a dumb use case (it arised to me as a temporary fix for a division by 0 which I should have avoided by other means anyway). Nonetheless - if at least for the sake of mathematical consistency - it would be nice to fix it to evalutate to 0 or raise at an exception that users will know they're doing something pointless.

I have some exemplary code that reproduces the behaviour (run on GPU with fastmath = True):

```
import numpy as np
import theano
import theano.tensor as T


W = theano.shared(np.asarray(np.random.normal(0, 0.1, size=(10,10)), dtype=np.float32), borrow=True, name='W')
x = T.fvector()
y = T.fvector()

out_ = T.dot(x,W)
error_ = T.square(out_ - y).sum()
error_ = error_ / 0
error_      = T.where(T.isnan(error_), 0, error_)
error_      = T.where(T.isinf(error_), 0, error_)
grad_  = T.grad(error_, W)

foo = theano.function([x,y], [out_, error_, grad_])
d = np.ones(10, dtype=np.float32)
l = np.ones(10, dtype=np.float32)
out, error, grad = foo(d,l)
grad = np.asarray(grad)
print error, grad
```

Btw, setting the error_ to zero by multiplying 0 works however:

```
out_ = T.dot(x,W)
error_ = T.square(out_ - y).sum()
error_ = error_ * 0
grad_  = T.grad(error_, W)

foo = theano.function([x,y], [out_, error_, grad_])
d = np.ones(10, dtype=np.float32)
l = np.ones(10, dtype=np.float32)
out, error, grad = foo(d,l)
grad = np.asarray(grad)
print error, grad
```

Cheers
Marius

**Comment by lamblin** [comment]
On Sat, Jun 06, 2015, xeray wrote:
> Hello,
> 
> I have observed that taking the gradient of a constant w.r.t to parameters gives NaN for the gradient of those parameters, where I expect it to be zero. I know, taking the gradient of a constant is sort of a dumb use case (it arised to me as a temporary fix for a division by 0 which I should have avoided by other means anyway). Nonetheless - if at least for the sake of mathematical consistency - it would be nice to fix it to evalutate to 0 or raise at an exception that users will know they're doing something pointless.
> 

fastmath will not work correctly with nan and inf, as it assumes that they will not occur.
In particular, "isinf" or "isnan" could return 0 even if the value is actually nan.

I just tried your example on CPU, without fastmath, and grad was full of zeros.

> I have some exemplary code that reproduces the behaviour (run on GPU with fastmath = True):
> 
> ```
> import numpy as np
> import theano
> import theano.tensor as T
> 
> 
> W = theano.shared(np.asarray(np.random.normal(0, 0.1, size=(10,10)), dtype=np.float32), borrow=True, name='W')
> x = T.fvector()
> y = T.fvector()
> 
> out_ = T.dot(x,W)
> error_ = T.square(out_ - y).sum()
> error_ = error_ / 0
> error_      = T.where(T.isnan(error_), 0, error_)
> error_      = T.where(T.isinf(error_), 0, error_)
> grad_  = T.grad(error_, W)
> 
> foo = theano.function([x,y], [out_, error_, grad_])
> d = np.ones(10, dtype=np.float32)
> l = np.ones(10, dtype=np.float32)
> out, error, grad = foo(d,l)
> grad = np.asarray(grad)
> print error, grad
> ```
> 
> Btw, setting the error_ to zero by multiplying 0 works however:
> 
> ```
> out_ = T.dot(x,W)
> error_ = T.square(out_ - y).sum()
> error_ = error_ * 0
> grad_  = T.grad(error_, W)
> 
> foo = theano.function([x,y], [out_, error_, grad_])
> d = np.ones(10, dtype=np.float32)
> l = np.ones(10, dtype=np.float32)
> out, error, grad = foo(d,l)
> grad = np.asarray(grad)
> print error, grad
> ```
> 
> Cheers
> Marius
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Theano/Theano/issues/3006

-- 
Pascal

**Comment by nouiz** [comment]
@lamblin answer seem right, so I'll close this issue. If you still have questions/comments, don't hesitate to continue the discussion.

**Comment by xeray** [comment]
Hi Pascal,

thanks for your explanation. So I understand that I have to either turn off fastmath or stabilise my code manually by adding epsilons to expressions that coult give rise to Nan/Inf.

Thanks
Marius

**Comment by xeray** [comment]
Hi Pascal and Frédéric,

it's been half a year now and I ran into the same issue again. This time I intended to use np.nan to mark targets that are not available (--> mask with ``T.isnan`` and set loss to ``0.0`` where applicable).

Unfortunately I didn't remember my own issue here - unless I spent quite some time trying to figure out where my devastating ``nan``-gradients originate from.

I think the current state of this issue is highly dangerous! From a plain logical point of view getting ``0.0`` instead of ``nan`` as gradient is so clear that one wouldn't even consider the possibility that this assumption gives rise to severe errors.

I guess the most common use case of theano is to run it on the GPU in conjunction with fastmath and if theano offers and implements the function ``isnan`` it must either be consistent or yell loud and furious warnings at the user that is not consistent, it should not just fail silently.

If you don't want to fix this I'm open for suggestions how to do it and what other functions might be affected and make fix+PR myself.

Cheers
Marius

**Comment by lamblin** [comment]
I don't think that Theano has to be stricter in that regard than the C or CUDA compiler, which do offer an `isnan` function that is inconsistent when `-ffastmath` is passed.
That being said, if you have an idea how the behaviour could be changed, we can discuss it.

**Comment by xeray** [comment]
Hi Pascal,

thanks for your fast reply.
Given that we cannot change the specs of the compilers, I think the best idea is to issue a warning (or even raise an exception) at the point when the ``isnan``-node is translated to C or CUDA code by the theano compilation pipeline (conditioned on the stat of the fastmath flag).

However, I am not sure whether there is not an additional problem independent of fastmath. Because actually in the forward pass I get the expected result even when using ``isnan`` in fastmath mode, and only the gradient gives ``nan``. This happens without fastmath or in python mode, too.

```
import numpy as np
import matplotlib.pyplot as plt
import theano
import theano.tensor as T

pred = T.fvector("input")
target = T.fvector("target")

mask = T.isnan(target)
out    = T.square(target - pred)

out = T.set_subtensor(out[mask.nonzero()], 0.0) # setting some values to a constant
loss = out.sum()
grad = theano.grad(loss, pred) # expect grad of a constant is 0

f = theano.function([pred, target], [out, grad])

pred_val   = np.array([0.0, 0.3, 0.4, 0.0, 0, 0.3], dtype=np.float32)
target_val = np.array([0.3, 0.0, 0.4, 0.0, np.nan, np.nan], dtype=np.float32)

tmp = f(pred_val, target_val)
print(tmp)
# [ 0.09,  0.09,  0.  ,  0.  ,  0.  ,  0.  ],[-0.6,  0.6, -0  , -0 , nan, nan]
```
I don't understand why this happens.

Cheers
Marius

**Comment by lamblin** [comment]
OK, so in that case it has nothing to do with fastmath.
The issue is the way the gradient of `square` is implemented, as `2 * dloss/dout * (target - pred)`. `dloss/dout` is `[ 1.,  1.,  1.,  1.,  0.,  0.]`, so at that point at the backprop, the gradient wrt the constant output is 0, but then it gets multiplied by `target - pred`, which is `[ 0.30000001, -0.30000001,  0.        ,  0.        ,         nan,                nan]`.

Handling `nan` differently in those gradients when the output gradient is 0 would be tricky in general, so it may be better for you to filter out the nans earlier if possible.

**Comment by xeray** [comment]
Ok, that makes sense to me. Thanks for explaining.
